### DISCUSSION TITLE
Convex optimization of an optimal control problem-error

### DISCUSSION THREAD

**Discussion by Sahil-717** [yalmip/YALMIP#1186] [Optimization modelling]
Dear Professor, 
For the following optimization problem (all parameters are are matrices with dimension 4), I face this error by using Mosek solver in MATLAB: 
**Model creation failed (Model not available in objective at level 1)**

I want to know if my problem, can be simulated as a QP problem by Mosek or I should redefine the problem specially the constraints related to the system dynamics.

![Capture3](https://user-images.githubusercontent.com/114741444/193261206-3863c41b-bc41-4f95-ba20-884faed7dd72.JPG)

**Comment by johanlofberg** [comment]
Mosek (or any convex solver for that matter) is not applicable as you have bilinears in the equalities

**Reply by johanlofberg** [reply]
...and time derivatives does not exist in YALMIP so it is not even clear what you are solving

**Reply by Sahil-717** [reply]
Dear Professor Löfberg,
Thank you for your kind reply. 
About bilinear statements in the objective function, is it possible to write them as follows,

![image](https://user-images.githubusercontent.com/114741444/193531201-ebda34cb-06b0-4d7c-8191-b9dc9a7779d6.png)

which implies a quadratic objective function?

Besides, using the system dynamics as constraints of the problem in YALMIP is inspired by the following example of MPC in "yalmip.github.io", 
![image](https://user-images.githubusercontent.com/114741444/193532783-6438ba1c-3781-40ff-8bde-8e15c1572c5e.png)
Can I implement the discrete system dynamics in the same way?

**Reply by johanlofberg** [reply]
I don't see the difference in the question. The problem I refered to was the bilinear elements in the equalities (which on the other hand defined a differential equation, which makes no sense as YALMIP has no notion of differential equations). The objective is problematic regardless, as it is appears convex quadratic but you are maximizing it, hence nonconvex problem.

**Reply by Sahil-717** [reply]
Dear Professor, thank you for your answer. What should I do with the bilinear state equations? Should I use convex relaxation methods in order to deal with these equation?

**Reply by johanlofberg** [reply]
Well, it is very unclear what you even intend to model as an optimization problem, as you currently have described a differential equation, so the objective makes no sense as there is no reference to time (is the cost intended to be an integral over time? there must be some initial condition? etc)

Also, it is not clear what actually is a decision variable, the "variable" Omega is a function of some other omegahat, and it is not clear if that is the decision variable (same with zeta, is v the decision variable?) Why is there an index i, and is there any connection between different i and if not why not simply concentrate on one of them as it appears to be completely separable

**Reply by Sahil-717** [reply]
In fact, the original problem is optimal control of the direction of rigid bodies (multi-agent system) as follows:
![Eq1](https://user-images.githubusercontent.com/114741444/225566839-70f3c8a7-80be-4751-b052-91eb5f3df44e.JPG)

The indices "i" and "j" represents the i-th  and j-th agents, with state-space equations as:
![image](https://user-images.githubusercontent.com/114741444/225569514-a24ef303-f744-4448-95f6-340e9b11d0e0.png)

while:  vi,ωi ∈ R^3 are respectively the linear and angular velocities of agent "i" with respect to the body frame., that should be designed in order to control the agents.

The original problem is rewritten to the following optimization problem on SO(3)
![image](https://user-images.githubusercontent.com/114741444/225569038-d4631612-39de-4862-94d8-fa73ad14d994.png)
while
![image](https://user-images.githubusercontent.com/114741444/225571309-fff8e338-4b63-4394-ad7c-f5594eeef228.png) ![image](https://user-images.githubusercontent.com/114741444/225572007-602c4e5b-fa5f-4028-81f1-292d5a94b0ee.png)  is a weighting function, with symmetric positive definite matrices
![image](https://user-images.githubusercontent.com/114741444/225571493-c17ab421-8654-45df-9a3a-d804d8c158b6.png)

Finally, the problem is convexifed to the following optimization problem
![image](https://user-images.githubusercontent.com/114741444/225572826-708e7649-bd36-4020-8751-06aedad48a72.png)

Is the problem stated in the correct way?

**Reply by Sahil-717** [reply]
Besides, I want to know if the optimization should be introduced just in terms of ![image](https://user-images.githubusercontent.com/114741444/225579141-8dd4928b-d5b5-4ab8-b8c3-c62ad8360487.png) or ![image](https://user-images.githubusercontent.com/114741444/225578802-aef4828c-ccc1-4dcb-845b-8ee1927ccdd1.png) also should be considered as the optimization variable?

I would be so appreciative if you help me.

**Reply by johanlofberg** [reply]
It's weird that minimization of a sum of quadratic stuff (good) can turn into maximization of quadratic stuff (bad, immediately nonconvex)

As I said, at the moment it is not a finite-dimensional optimization problem that can be addressed using YALMIP (or any related solver) as it is an continuous-time optimal control problem in omegahat and v. There is no thing such as time or time-derivative in a standard optimization problem.

**Reply by Sahil-717** [reply]
Changing from min to max is because of the convexification of the problem related to the term ![Eq3](https://user-images.githubusercontent.com/114741444/226063625-ff4f983a-6411-430a-b5d6-cd757d324c6e.JPG), that a minus sign rise up.

**Reply by Sahil-717** [reply]
Related to considering state-space equations as the constraints of the optimization problem, I want to know if it is not possible to consider the time derivative of the equations as equality constraints. The same as the following code
![Eq4](https://user-images.githubusercontent.com/114741444/226065233-fe3d6f94-52a1-4d19-8cce-be6ca2fda398.JPG)
which considers the state equations for updating the objectives.

**Reply by johanlofberg** [reply]
I repeat, `minimize <W*X*X> `is convex, while `maxmize <W*X,X>` is nonconvex. Hence, you are going from something which looks convex, to something which definitely is nonconvex, and it sounds very strange that any kind of convexification would convert a positive definite quadratic term to a negative definite term (since any you have inner products, any sign-switch in a term in a thus formmed quadratic would disappear anyway and render the term untouched)

The MPC examples are discrete-time models, you have a continuous-time model.
